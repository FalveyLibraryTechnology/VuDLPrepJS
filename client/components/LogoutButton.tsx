import React, { useEffect, useState } from "react";

import { logoutUrl } from "../util/routes";
import { useFetchContext } from "../context/FetchContext";

const LogoutButton = (): React.ReactElement => {
    const {
        state: { token },
        action: { clearToken },
    } = useFetchContext();
    const [showButton, setShowButton] = useState<boolean>(false);
    useEffect(() => {
        setShowButton(token !== null);
    }, [token]);
    return showButton ? (
        <div className="logout">
            <a href={logoutUrl} className="button btn-primary" onClick={clearToken}>
                Log Out
            </a>
        </div>
    ) : (
        <></>
    );
};

export default LogoutButton;
