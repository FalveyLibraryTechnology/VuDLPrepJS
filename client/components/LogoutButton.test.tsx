import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import LogoutButton from "./LogoutButton";

const mockUseFetchContext = jest.fn();
jest.mock("../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));

describe("LogoutButton", () => {
    let fetchValues;
    beforeEach(() => {
        fetchValues = {
            state: {
                token: "foo",
            },
            action: {
                clearToken: jest.fn(),
            },
        };
        mockUseFetchContext.mockReturnValue(fetchValues);
    });

    it("renders correctly when logged in", async () => {
        let tree;
        await renderer.act(async () => {
            tree = renderer.create(<LogoutButton />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("renders correctly when logged out", async () => {
        fetchValues.state.token = null;
        let tree;
        await renderer.act(async () => {
            tree = renderer.create(<LogoutButton />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("clears the token", async () => {
        render(<LogoutButton />);

        expect(fetchValues.action.clearToken).not.toHaveBeenCalled();

        // Clicking a link will trigger a console error about navigation not being
        // implemented -- but that doesn't hurt the test, so let's suppress it by
        // mocking the error method.
        // TODO: figure out why and come up with a better solution than hiding the errors.
        jest.spyOn(console, "error").mockImplementation(jest.fn());
        await userEvent.setup().click(screen.getByText("Log Out"));
        expect(fetchValues.action.clearToken).toHaveBeenCalled();
    });
});
