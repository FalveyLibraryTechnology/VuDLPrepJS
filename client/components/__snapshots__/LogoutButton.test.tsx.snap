// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutButton renders correctly when logged in 1`] = `
<div
  className="logout"
>
  <a
    className="button"
    href="http://localhost:9000/api/auth/logout"
    onClick={[MockFunction]}
  >
    Log Out
  </a>
</div>
`;

exports[`LogoutButton renders correctly when logged out 1`] = `null`;
